serving/purgatorio/select-and-train-machine-learning-models/tools-and-libraries/tools-and-libraries.md

```diff
@@ -1,5 +1,321 @@
+TF 2.0 is the tools chosen in the up-to-date version of the mythical (and _"mandatory"_) book [Hands-On Machine Learning with Scikit-Learn, Keras, and Tensorflow](https://www.amazon.it/Hands-Machine-Learning-Scikit-learn-Tensorflow/dp/1492032646).
+
+Bonus tip:
+
+Don't try to learn the old TF1 version, hadn't aged well, but Google learned a lot from that experience: go for 2.0.
+
+- [TensorFlow 2.0: Easier for beginners, and more powerful for experts](https://www.youtube.com/watch?v=5ECD8J3dvDQ)
+
+---
+
+To learn TensorFlow you'll need mostly a lot of practice, starting from toy examples to most complex projects, but Virgilio collected for you the most important resources out there:
+
+- [Here](https://www.youtube.com/watch?v=tYYVSEHq-io&t=2s) you have a gentle and brilliant introduction to what is TensorFlow.
+- [Here](https://www.tensorflow.org/tutorials) you find the official Tutorials.
+- [Here](https://www.tensorflow.org/guide) you find the official Guide, a more detailed guide to each aspect of the framework.
+- [Here](https://github.com/jtoy/awesome-tensorflow) you have an awesome list of additional resources about Tensorflow.
+
+Bonus tip: 
+
+There is a reason why Google engineers decided to go integrate Keras natively in TF 2.0: because it's awesome!
+Most of the time you don't need to dive **so deep** inside a neural network architecture to make it work!
+
+- [You should start with Keras](https://www.reddit.com/r/learnmachinelearning/comments/biynuy/difference_between_keras_and_tensorflow/)
+
+You should be able to find different frameworks and dozens of other exotic neural network libraries on the Internet, but as usual, _Virgilio highly recommends to use your time wisely_ (one enemy at the time). 
+It's not always true that the most used thing is the best, but for sure you can be certain that Tensorflow is here to stay and to remain one (ore the most) widely used deep learning framework.  
+
+## Hyperparameter Tuning
+
+If you’ve ever tried to tune hyperparameters for a machine learning model, you know that it can be a very painful process!
+
+To fine-tune your models you can check a variety of tools, check this article:
+
+- [A Comprehensive List of Hyperparameter Optimization & Tuning Solutions](https://medium.com/@mikkokotila/a-comprehensive-list-of-hyperparameter-optimization-tuning-solutions-88e067f19d9)
+
+In particular, if you are learning [Keras](https://www.tensorflow.org/guide/keras) as a top-level framework on the top of Tensorflow, Virgilio suggests you check [Hyperas](https://github.com/maxpumperla/hyperas).
+
+A very simple convenience wrapper around [Hyperopt](https://github.com/hyperopt/hyperopt) for fast prototyping with Keras models. Hyperas lets you use the power of hyperopt without having to learn the syntax of it. Instead, just define your Keras model as you are used to, but use a simple template notation to define hyper-parameter ranges to tune.
+
```

Comment:
There are related tools for hyper-parameter tuning, see this search result:
https://github.com/neomatrix369/awesome-ai-ml-dl/search?q=hyper&unscoped_q=hyper

Reply:
yes I wanted to pick one to not overwhelm the reader 👍